src/crawlspace/handlers/external.py

```diff
@@ -1,50 +1,177 @@
 """Handlers for the app's external root, ``/crawlspace/``."""
 
-from fastapi import APIRouter, Depends
+from dataclasses import dataclass
+from email.utils import format_datetime
+from mimetypes import guess_type
+from typing import Dict, Iterator, List
+
+from fastapi import APIRouter, Depends, HTTPException, Path, Request, Response
+from fastapi.responses import RedirectResponse, StreamingResponse
+from google.cloud import storage
 from safir.dependencies.logger import logger_dependency
-from safir.metadata import get_metadata
 from structlog.stdlib import BoundLogger
 
 from ..config import config
-from ..models import Index
-
-__all__ = ["get_index", "external_router"]
+from ..constants import CACHE_MAX_AGE
+from ..dependencies.caching import cache_validation_dependency
+from ..dependencies.gcs import gcs_client_dependency
 
 external_router = APIRouter()
 """FastAPI router for all external handlers."""
 
+_PATH_REGEX = r"^(([^/.]+/)*[^/.]+(\.[^/.]+)?)?$"
+"""Regex matching a valid path.
+
+Path must either be empty, or consist of zero or more directory names that
+do not contain ``.``, followed by a file name that does not contain ``.``
+and an optional simple extension introduced by ``.``.
+
+This is much more restrictive than the full POSIX path semantics in an attempt
+to filter out weird paths that may cause problems (such as reading files
+outside the intended tree) when used on POSIX file systems.  This shouldn't be
+a problem for GCS, but odd paths shouldn't be supported on GCS anyway.
+"""
```

Comment:
I wonder if you could instead test for paths outside the tree by turning the path into a pathlib.PurePosixPath and then checking with the `is_relative_to` method? Not really a suggestion to do, but sometimes regexes like these give me pause.

Reply:
Out-of-tree files is only an example. I don't really want people to use weird Unicode characters like bidirectional switching, nul characters, or whatever hypothetical characters might be treated differently by Google Cloud Storage either. From a security perspective I prefer to pass only the files I want to allow rather than trying to detect malicious input.

I agree that the regex is kind of ugly and hard to understand, which is unfortunate. Not sure if there's a better validation approach.